jax/lax/lax_partition.py

```diff
@@ -0,0 +1,312 @@
+
+
+def _partition_translation_rule(
+    c, axis_env, name_stack, *in_nodes, jaxpr, partition_ids, backend):
+  jaxpr = jaxpr.jaxpr  # The `jaxpr` parameter is a `TypedJaxpr`.
+  nodes = {}
+
+  def read_node(v):
+    if type(v) is core.Literal:
+      return c.Constant(xla.canonicalize_dtype(v.val))
+    else:
+      return nodes[v]
+
+  def write_node(v, node):
+    assert node is not None
+    nodes[v] = node
+
+  write_node(core.unitvar, c.Tuple())
+  _map(write_node, jaxpr.invars, in_nodes)
+
+  var_partition_ids = partition_util.get_var_partition_ids(jaxpr, partition_ids)
+
+  for eqn in jaxpr.eqns:
+    invars = [v for v in eqn.invars if type(v) is not core.Literal]
+
+    if invars:
+      # Take partition ID from the first input var. We've already checked, in
+      # `_get_var_partition_ids`, that they all match.
+      partition_id = var_partition_ids[invars[0]]
+    else:
+      partition_id = None
+
+    with _op_partition(c, partition_id):
+      out_nodes = xla.jaxpr_subcomp_eqn(
+          c, eqn, backend, axis_env, name_stack, read_node)
+    _map(write_node, eqn.outvars, out_nodes)
+  return c.Tuple(*_map(read_node, jaxpr.outvars))
+
+
+def _partition_jvp(primals, tangents, jaxpr):
+  raise NotImplementedError()  # TODO(cjfj)
+
+
+partition_p = core.Primitive('partition')
+partition_p.multiple_results = True
+partition_p.def_impl(_partition_impl)
+partition_p.def_abstract_eval(_partition_abstract_eval)
+ad.primitive_jvps[partition_p] = _partition_jvp
+xla.initial_style_translations[partition_p] = _partition_translation_rule
+
+
+def partition_put(x, partition_id):
+  """Places the given value(s) onto the specified partition.
+
+  Note that for nested `partition` calls, the `partition_id` is the index into
+  the list of partition IDs used by the innermost partitioned function.
+
+  Example:
+    ```
+    @partial(partition, num_partitions=2)
+    def f(w):
+      x = w ** 2
+      y = jax.partition_put(x, 1)
+      z = 42 * y
+      return z
+
+    @partial(partition, num_partitions=4)
+    def g(i):
+      j = i ** 3
+      k = f(i, outer_partition_ids=[1, 2])
+      l = 42 * jax.partition_put(k, 3)
+      return l
+
+    b = g(a)
+    ```
+
+    `g` is the outermost partitioned function here, so partition IDs map
+    directly to the logical XLA devices. `f` is nested, with its two partitions
+    mapped onto `[1, 2]` using `outer_partition_ids`. `x` is therefore
+    calculated on XLA partition 1, whilst `z` is calculated on XLA partition 2.
+
+  Args:
+    x: The value(s) to copy. This can be a tree of values.
+    partition_id: The partition ID, within the innermost `partition` scope,
+      on which to place the value(s).
+
+  Returns:
+    The input values, copied onto the specified partition.
+  """
+  return tree_map(partial(partition_put_p.bind, partition_id=partition_id), x)
+
+
+def _partition_put_impl(x, *, partition_id):
+  del x, partition_id  # Unused.
+  raise NotImplementedError(
```

Comment:
nit: make this a ValueError or something, since presumably we'll never implement this?

Reply:
Done